core/src/main/java/hudson/model/Descriptor.java

```diff
@@ -918,17 +918,29 @@ private URL getStaticHelpUrl(Klass<?> c, String suffix) {
                 // Descriptors, so we prefer 'kind' if it's present.
                 String kind = jo.optString("kind", null);
                 if (kind != null) {
+                    // Only applies when Descriptor.getId is overridden.
+                    // Note that kind is only supported here,
+                    // *not* inside the StaplerRequest.bindJSON which is normally called by newInstance
+                    // (since Descriptor.newInstance is not itself available to Stapler).
+                    // If you merely override getId for some reason, but use @DataBoundConstructor on your Describable,
+                    // there is no problem; but you can only rely on newInstance being called at top level.
                     d = findById(descriptors, kind);
                 }
                 if (d == null) {
-                  kind = jo.getString("$class");
-                  d = findByDescribableClassName(descriptors, kind);
-                  if (d == null) d = findByClassName(descriptors, kind);
+                  kind = jo.optString("$class");
+                  if (kind != null) { // else we will fall through to the warning
+                      // This is the normal case.
+                      d = findByDescribableClassName(descriptors, kind);
+                      if (d == null) {
+                          // Deprecated system where stapler-class was the Descriptor class name (rather than Describable class name).
+                          d = findByClassName(descriptors, kind);
+                      }
+                  }
                 }
                 if (d != null) {
                     items.add(d.newInstance(req, jo));
                 } else {
-                    LOGGER.warning("Received unexpected formData for descriptor " + kind);
+                    LOGGER.log(Level.WARNING, "Received unexpected form data element: {0}", jo);
```

Comment:
Yes, that is another possible failure mode: somehow the configuration form gets generated with a particular descriptor loaded and presented as an option, yet for some reason the descriptor cannot be located during form submission. (The only case I can think of that would not involve some weird bug would be that you create a builder template using `cloudbees-template`, configure a project, select the builder, delete the template in another browser tab, then submit the configuration page.)

At any rate, even in that case this patch does not lead to less diagnostic information; it just means that instead of

    Received unexpected formData for descriptor some.missing.Describable

you will see

    Received unexpected form data element: {"$class":"some.missing.Describable",…}

In the case I actually encountered that prompted this commit, the problem was not a missing `Descriptor` but a `JSONObject` with the wrong structure, so displaying the full object would be more helpful. Also remember that it is possible for `kind == null`, in case the object contains neither `kind` nor `$class`, in which case dumping the full object might give a clue what is going on.